public/dorequest.php

```diff
@@ -277,6 +337,78 @@ function DoRequestError(string $error, ?int $status = 200, ?string $code = null)
         $response['AchievementID'] = $achIDToAward;
         break;
 
+    // This is only currently supported for "Standalone" integrations.
+    case "awardachievements":
+        if (request()->method() !== 'POST') {
+            return DoRequestError('Access denied.', 403, 'access_denied');
+        }
+
+        $achievementIdsInput = request()->input('a', '');
+        $validationHashesInput = request()->input('v', '');
+        $hardcore = (bool) request()->input('h', 0);
+
+        $foundTargetUser = null;
+        if (!$delegateTo) {
+            return DoRequestError("You must specify a target user.", 400);
+        }
+
+        $foundTargetUser = User::firstWhere('User', $delegateTo);
+        if (!$foundTargetUser) {
+            return DoRequestError("The target user couldn't be found.", 400);
+        }
+
+        $achievementIdsArray = explode(',', $achievementIdsInput);
+        $validationHashesArray = explode(',', $validationHashesInput);
+
+        if (count($achievementIdsArray) !== count($validationHashesArray)) {
+            return DoRequestError("The number of achievements and hashes must match.", 400);
+        }
+
+        $idsToAttemptAward = array_filter($achievementIdsArray, function ($id) {
+            return filter_var($id, FILTER_VALIDATE_INT) !== false;
+        });
+
+        // Fetch the IDs of achievements already awarded to the user.
+        $alreadyAwardedIds = PlayerAchievement::where('user_id', $foundTargetUser->id)
+            ->whereIn('achievement_id', $idsToAttemptAward)
+            ->pluck('achievement_id')
+            ->all();
```

Comment:
I'm still uncertain how standalones will handle hardcore vs. softcore, but this logic will not acknowledge an achievement being unlocked in hardcore if it has previously unlocked in softcore.

Reply:
I've updated the code to account for this. It does add a bit more complexity to the `awardachievements` logic.